Stack deploy improvements

nikoglyph opened the pull request:
Hi Tobias,

While looking, in vain as it turned out, for docker engine API support for docker stack operations, in relation to a project at work, I stumbled across your client library, and have now been exploring it over the weekend.

I'm very impressed with your work - the organisation and clarity of the code, the extent of the API-coverage, the test coverage, and not least, your well-reasoned argument for implementing the client in a dynamically typed language.

To meet our requirements, I've made one minor change, fixed a bug, and slightly revised some tests to make them succeed locally.

An overview of the changes in the pull request, in order of importance:

1) Added a second loadCompose method to DeployConfigReader, to allow client code to directly pass in an environments Map (and thus avoiding polluting the environment of the client host/jvm in the process). This option lacks from the 'docker stack deploy' cli command, but it could be argued that this is an oversight, as it is possible to pass environment variables when creating services with 'docker service create --env').

2) Fixed a bug in the conversion of deploy.mode: global.

3) Revised integration test for import image from url, to not rely on the first network interface address (on my machine the first address was for a docker interface).

4) Revised integration test for 'docker version' to expect only a certain set of keys, not their values. In the wider perspective, as your client library finds more contributors, the expectations of integration tests should probably be more modest, to allow for some diversity in development platforms/docker versions.

5) Revised integration test for distribution image descriptor, in the same vein.

6) Added (starting point for) integration test for 'docker stack deploy -c'. The idea was to have more of an end-to-end test, with docker-stack.yml inlined in the test class to make it easy to see what is verified, and to expand it with as many supported compose options/keys as possible. I don't know if this is the right approach, maybe to start with because the test will grow large. Anyway, a means in some form of verifying that all compose file stack options, fit together in a complete stack description, produces the expected deployed services, networks and volumes, etc, would be useful.

7) Added gradle install task (maven plugin), for installing SNAPSHOT version of client library to local maven repo - purely a convenience in our setting, for fast turn-around with locally built client library used by our maven-based application.   


Best regards,
Nicholas

gesellix commented:
Hi Nicholas,

well, thanks a lot for such a beauty of a pull request! Your PR enlightened my day and if there would be a price for the best/most beautiful/nice/concise PR of the year, this one would be my favourite!

I'll still have to look into the actual changes, yet there's one thing I would already propose to change after reading your introduction. You mentioned in 6. that you used inline YAML snippets for the new integration test. Would you mind to move the snippets into dedicated `.yml` files and read those files as part of the test cases? Just like you already mentioned I wouldn't like the test class to be cluttered with noisy amount of YAML snippets. Please also note that you should try to focus on testing the docker-client side of the end, because the actual `.yml` reader should be tested in its dedicated library project at https://github.com/docker-client/docker-compose-v3. The idea is simple: `docker-compose-v3` reads YAML and yields Java objects. The `docker-client` takes the compose/stack specific Java objects and converts them to its own types/format. That said: if there are some overlapping tests, that's ok for me, so please don't think you should delete everything... some more overlapping tests are better than bad test coverage ;-)

I will have a more detailed look into it tomorrow. Again: thanks a lot for your huge efforts!

nikoglyph commented:
Thanks! I'll take a look at the integration test tonight, with your feedback in mind!

gesellix commented on integration-test/src/test/groovy/de/gesellix/docker/client/DockerStackComposeIntegrationSpec.groovy:
```diff
@@ -0,0 +1,162 @@
+        performSilently { dockerClient.leaveSwarm([force: true]) }
+    }
+
+    def setupSpec() {
```
Please have a look at the other specs, I prefer an order like this:

- global fields for every test case
- setup method(s)
- cleanup method(s)
- test cases
- util methods/fixture factories

nikoglyph replied:
I've moved the inlined YAML to a resource file, reordered fields/methods according to the above, and cleaned-up a bit. A future refactoring could be to move utility methods for delaying/retrying docker command-closures out to a separate utilities class - if those are useful also in other integration tests.

gesellix commented on integration-test/src/test/groovy/de/gesellix/docker/client/DockerStackComposeIntegrationSpec.groovy:
```diff
@@ -0,0 +1,162 @@
+
+    def "deploy a new stack with compose file"() {
+        given:
+        dockerClient.initSwarm(newSwarmConfig())
```
There's an overloaded `initSwarm()` method available without any parameters. It internally initializes a default swarm config. Feel free to change to the simpler form, then you can remove the `newSwarmConfig()` method in this test. I'm aware that the other integration tests haven't been cleaned accordingly, yet ;-)

nikoglyph replied:
Done

nikoglyph commented on client/src/main/groovy/de/gesellix/docker/client/stack/DeployConfigReader.groovy:
```diff
@@ -52,7 +52,11 @@ class DeployConfigReader {
 
     // TODO test me
     def loadCompose(String namespace, InputStream composeFile, String workingDir) {
```
Done

nikoglyph commented on integration-test/src/test/groovy/de/gesellix/docker/client/DockerDistributionIntegrationSpec.groovy:
```diff
@@ -22,18 +22,11 @@ class DockerDistributionIntegrationSpec extends Specification {
-                        ]
-                ]
+        alpineDescriptor.content.Descriptor.mediaType == "application/vnd.docker.distribution.manifest.list.v2+json"
+        alpineDescriptor.content.Descriptor.digest != null
```
Good point, done

nikoglyph commented on integration-test/src/test/groovy/de/gesellix/docker/client/DockerDistributionIntegrationSpec.groovy:
```diff
@@ -44,21 +37,17 @@ class DockerDistributionIntegrationSpec extends Specification {
-                        [architecture: "s390x", os: "linux"]
-                ]
+        debianDescriptor.content.Descriptor.mediaType == "application/vnd.docker.distribution.manifest.list.v2+json"
+        debianDescriptor.content.Descriptor.digest != null
```
Done

gesellix commented:
Looks good, travis-ci has passed, too. Great job!
Can you rebase so that I can merge? If you like you can also squash your commits, but that's not mandatory.

nikoglyph commented:
Ok, now I'm done - hopefully travis agrees. I've squashed my work into a single commit and rebased. By the way, your constructive code review was much appreciated - and thanks for your great work with this library!

gesellix commented:
Thanks a lot!

gesellix commented:
The new release is available as version `2017-10-11T07-03-57`.

nikoglyph commented:
Ok, thanks, will do.
